toolchain/docs/parse.md

```diff
@@ -0,0 +1,917 @@
+    ```
+
+    -   `state.subtree_start` is the value of `context.tree().size()` saved in
+        step 1, which marks the position of the placeholder node in the output
+        parse tree.
+
+    -   `introducer_kind` is the `NodeKind` for the introducer of this
+        declaration or statement, a leaf node that will act as a bracketing node
+        at the beginning of the subtree for this declaration or statement
+
+-   **Step 4 - error**: If we run into something other than a modifier or
+    introducer before finding an introducer, we need to do error handling:
+
+    ```cpp
+    context.ReplacePlaceholderNode(subtree_start, NodeKind::InvalidParseStart,
+                                   *context.position(), /*has_error=*/true);
+    ```
+
+    -   Emit a [diagnostic](diagnostics.md).
+
+    -   Replace the placeholder node (similar to step 4) with an
+        `InvalidParseStart` node. It will be associated with the unexpected
+        token that triggered this error.
+
+    -   Consume input token up to the likely end of the end of the current
+        statement or declaration. For example, we might consume up to a `;` or a
+        token at a lesser indent level using `context.SkipPastLikelyEnd(...)`.
+        It is important that we consume at least one token in the error case,
+        otherwise we could have an infinite loop of generating the same error on
+        the same token.
+
+    -   Emit a `InvalidParseSubtree` node. This will be the parent of any
+        emitted modifier nodes, and will be bracketed by the `InvalidParseStart`
+        node emitted above. It should be associated with the last token
+        consumed.
+
+        ```cpp
+        // Set `iter` to the last token consumed, one before the current position.
+        auto iter = context.position();
+        --iter;
+        context.AddNode(NodeKind::InvalidParseSubtree, *iter, subtree_start,
+                        /*has_error=*/true);
+        ```
+
+-   **Step 5**: (If success at step 4) Push whatever states are to be used to
+    parse the rest of the declaration. The first state pushed (the last state to
+    be processed) will handle the end of this declaration. That pushed state
+    should have a `subtree_start` field set to the value of
+    `context.tree().size()` saved in step 1.
+
+-   **Step 6**: When handling the state for the end of the declaration, emit the
+    root node of subtree:
+
+    ```cpp
+    state = context.PopState();
+    context.AddNode(NodeKind::___, context.Consume(),
+                    state.subtree_start, state.has_error);
+    ```
+
+    -   This `state.subtree_start` will mark everything since the bracketing
+        introducer node as the children of this node.
+
+## Something required in context
+
+FIXME
+
+Example: name after introducer
+[parse/handle_decl_name_and_params.cpp](/toolchain/parse/handle_decl_name_and_params.cpp)
+
+Example: `[<implicit parameter list>]` after `impl forall`
```

Comment:
```suggestion
Example: `[` <implicit parameter list> `]` after `impl forall`
```

Reply:
`<>` are HTML characters, so I was choosing to make this code formatted instead of `&lt;` which seemed less readable for people who look at the md file instead of the HTML